Fix GasMeter reset in AnteHandler - EthGasConsumeDecorator

loredanacirstea opened the pull request:
Conforming to the spec:
https://github.com/cosmos/cosmos-sdk/blob/e7066c4271ff3d33dc426dc6313c82a1201ae3c6/docs/basics/gas-fees.md#antehandler
> AnteHandler
> Set newCtx.GasMeter to 0, with a limit of GasWanted.
> This step is extremely important, as it not only makes sure the transaction cannot consume infinite gas,
> but also that ctx.GasMeter is reset in-between each DeliverTx
> (ctx is set to newCtx after anteHandler is run, and the anteHandler is run each time DeliverTx is called).

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Fixes https://github.com/tharsis/evmos/issues/311

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

`EthSetupContextDecorator` ante handler initializes the gas meter as an infinite gas meter (with gas limit 0) - https://github.com/tharsis/ethermint/blob/7b22a535569f558dc9083e0833915c745c68be43/app/ante/eth.go#L463

This gas meter needs to be reset with the correct gas limit, which accounts for all messages in the transaction. Because then, this value is used to calculate the block's consumed gas, when it is prepared. See:

https://github.com/cosmos/cosmos-sdk/blob/2646b474c7beb0c93d4fafd395ef345f41afc251/baseapp/baseapp.go#L658-L659

https://github.com/tendermint/tendermint/blob/abdf717761e71dca53389a9d218fce8a84f43991/internal/mempool/mempool.go#L365

What was happening was `gasWanted` was always `0` (infinite gas meter has a limit of 0), so the block that was proposed would contain all the transactions from the mempool found at that time. When the block was applied, at DeliverTx time, all transactions over the block gas limit were not applied (which is correct), but they would have no way to return to the mempool, because they were already processed. Trying to get a transaction receipt for them would just fail with `transaction not found`.

See these logs: https://gist.github.com/loredanacirstea/a935737b2b1cf3cbc0e679efa1787cd0#file-evmos_logs2-txt-L1464

- We see receipts not found because the block has not been finalized yet: https://gist.github.com/loredanacirstea/a935737b2b1cf3cbc0e679efa1787cd0#file-evmos_logs2-txt-L759, which is normal.
- Block is finalized with 50! transactions: https://gist.github.com/loredanacirstea/a935737b2b1cf3cbc0e679efa1787cd0#file-evmos_logs2-txt-L908
- Immediately after the block is being processed and transactions applied to the state. Now we see a correct calculation of block gas, with each transaction.
- Block gas limit is reached at https://gist.github.com/loredanacirstea/a935737b2b1cf3cbc0e679efa1787cd0#file-evmos_logs2-txt-L1197 and the rest of the transactions included in the block are not processed.
- Receipts for the processed tx are returned https://gist.github.com/loredanacirstea/a935737b2b1cf3cbc0e679efa1787cd0#file-evmos_logs2-txt-L1245, but for the ones that are over the block gas limit (unprocessed), eth_getTransactionReceipt fails https://gist.github.com/loredanacirstea/a935737b2b1cf3cbc0e679efa1787cd0#file-evmos_logs2-txt-L1270.
- Next block is finalized with no data: https://gist.github.com/loredanacirstea/a935737b2b1cf3cbc0e679efa1787cd0#file-evmos_logs2-txt-L1464

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

fedekunze commented on app/ante/eth.go:
```diff
@@ -213,8 +218,11 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
 	}
 
+	// Set newCtx.GasMeter to 0, with a limit of GasWanted (gasLimit)
+	newCtx = ctx.WithGasMeter(sdk.NewGasMeter(gasTx.GetGas()))
```
I'd add the sum the gas values in the form loop above as there could be a potential case where a multi msg tx is crafted in a way that the tx gas is less than the sum of the msgs gas limits

loredanacirstea replied:
Done.

fedekunze reviewed (approved):
ACK, can you add a changelog.md entry under bug fixes? 🙏

fedekunze commented:
cc: @yihuang @thomas-nguy @tomtau @crypto-facs

yihuang commented on app/ante/eth.go:
```diff
@@ -213,8 +215,11 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
 	}
 
+	// Set newCtx.GasMeter to 0, with a limit of GasWanted (gasLimit)
+	newCtx = ctx.WithGasMeter(sdk.NewGasMeter(gasWanted))
```
I think the eth contract call's gas consumption is constrained by the EVM call itself? is it to constraint the gas consumed by cosmos-sdk operations?

loredanacirstea replied:
@yihuang I updated the PR description with more info. There was no way to cap transaction inclusion in the block, because the transaction gas limit was always 0. So, the constraint is for the cosmos-sdk operations.

yihuang replied:
I see, but the gas limit set for eth tx is not for the cosmos-sdk operations.
If we only need to solve tx inclusion issue, maybe it's safer to do this only for checkTx mode?

yihuang replied:
> I see, but the gas limit set for eth tx is not for the cosmos-sdk operations. If we only need to solve tx inclusion issue, maybe it's safer to do this only for checkTx mode?

or only at the end of ante handler in the last handler?

thomas-nguy replied:
I think the evm module will consume gas for all operations like saving variable in state or etc. not only EVM operations.  

May have a risk of mismatch?

yihuang replied:
I think we can set this gas meter in ante handler, but override with a infinite one in msg handler?

loredanacirstea replied:
@yihuang - "maybe it's safer to do this only for checkTx mode" - so do the gas meter replacement only if `ctx.IsCheckTx()` is true.

yihuang replied:
> @yihuang - "maybe it's safer to do this only for checkTx mode" - so do the gas meter replacement only if `ctx.IsCheckTx()` is true.

I was wrong about that, we need to report correct gasWanted in deliver mode too, but we only need to do that in ante handlers, so it's safe to override with an infinite one in msg handler: `Keeper.EthereumTx`, so I suggest to move the `NewInfiniteGasMeter` from `EthSetupContextDecorator` to `Keeper.EthereumTx`.

loredanacirstea commented on app/ante/eth.go:
```diff
@@ -182,6 +183,7 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
 		if err != nil {
 			return ctx, sdkerrors.Wrap(err, "failed to unpack tx data")
 		}
+		gasWanted += txData.GetGas()
```
`GetGas()` returns the amount of gas `wanted` (the equivalent of `gasLimit` in Ethereum) - https://github.com/cosmos/cosmos-sdk/blob/5d86db3ce593bcc8c8694f7f83418925f7e9568a/types/tx/types.go#L128-L129

As per the spec linked in the description: `the anteHandler returns a maximum limit of gas the transaction can consume during execution called GasWanted. The actual amount consumed in the end is denominated GasUsed`

thomas-nguy commented:
The gas meter is setup in `EthSetupContextDecorator` , with this approach it becomes redundant

fedekunze commented:
@thomas-nguy @yihuang @crypto-facs merging this

fedekunze commented:
@loredanacirstea tests are failing

yihuang commented on app/ante/eth.go:
```diff
@@ -213,6 +215,11 @@ func (egcd EthGasConsumeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simula
 	}
 
+	// Set newCtx.GasMeter to 0, with a limit of GasWanted (gasLimit)
+	if ctx.IsCheckTx() {
```
Sorry for the confusion from my comments, but we should do it for deliverTx too, but not for msg handler, so we should remove this condition, but override with a infinite gas meter at the beginning of `Keeper.EthereumTx` method.

loredanacirstea replied:
@yihuang , here:
https://github.com/tharsis/ethermint/blob/caa1c5a6c6b7ed8ba4aaf6e0b0848f6be5ba6668/x/evm/keeper/msg_server.go#L24-L26, right?

yihuang replied:
yes

yihuang reviewed (commented):
Sorry for the confusion from my comments, but we should do it for deliverTx too, but not for the msg handler, so we should remove the checkTx condition, but override with an infinite gas meter at the beginning of Keeper.EthereumTx method.

fedekunze reviewed (changes requested):
tests are failing

loredanacirstea commented:
The `TestOutOfGasWhenDeployContract` test is failing because it sets a gas meter with a small gas limit, which gets reset with the infinite gas meter from `Keeper. EthereumTx`. 
https://github.com/tharsis/ethermint/blob/caa1c5a6c6b7ed8ba4aaf6e0b0848f6be5ba6668/x/evm/handler_test.go#L409
@yihuang  Should the test be rewritten using some other API or is it a cause for concern?

The rpc tests are failing with the same errors as before. I will take a look in the morning.

peskypunk commented:
in case it helps, i am also seeing failures due to some sort of gas metering issue, which i tried to write up here:
https://gist.github.com/peskypunk/8c344c2028dbd319b6a4771e786cfa09

this is coming from the state of ethermint with @loredanacirstea's first 4 commits (which was the state of this PR until this morning)

yihuang commented:
> The `TestOutOfGasWhenDeployContract` test is failing because it sets a gas meter with a small gas limit, which gets reset with the infinite gas meter from `Keeper. EthereumTx`.
> 
> https://github.com/tharsis/ethermint/blob/caa1c5a6c6b7ed8ba4aaf6e0b0848f6be5ba6668/x/evm/handler_test.go#L409
> 
> 
> @yihuang Should the test be rewritten using some other API or is it a cause for concern?
> The rpc tests are failing with the same errors as before. I will take a look in the morning.

I think it should not expect a panic, but a `OutOfGas` error returned from here: https://github.com/tharsis/ethermint/blob/main/x/evm/keeper/state_transition.go#L383, the error should be able to check in the response.

loredanacirstea commented:
> I think it should not expect a panic, but a OutOfGas error returned from here: https://github.com/tharsis/ethermint/blob/main/x/evm/keeper/state_transition.go#L383, the error should be able to check in the response, similar to how the VM error is checked in test case TestErrorWhenDeployContract.

@yihuang , fixed the test here: https://github.com/tharsis/ethermint/pull/964/commits/892fc66815160d4533bd7165a3e55bba4db08366
Turns out that the out of gas error manifests as an `ErrIntrinsicGas` error from 
https://github.com/tharsis/ethermint/blob/f21592ebfe74da7590eb42ed926dae970b2a9a3f/x/evm/keeper/state_transition.go#L365-L369

RPC tests pass now: I moved the gas meter reset in the innermost EthAnteHandler. Previously it was failing because there was not enough gas for cosmos SDK operations (reading balances, creating accounts) that were done after the `EthGasConsumeDecorator`.

Note: First I tried adding another decorator after `EthIncrementSenderSequenceDecorator` with a more appropriate name, but the code was not being executed, even though I added it to `newEthAnteHandler` in `ante/handler_options.go`). So, then I just used `EthIncrementSenderSequenceDecorator`

loredanacirstea commented:
Do not merge yet. I just saw another problem, even though tests pass.

loredanacirstea commented:
@fedekunze , @yihuang 
PR ready. The cleanest way to fix the issue was a new type of gas meter that keeps an unenforced gas limit. 

(Resetting the gas meter in Keeper.EthereumTx did not work, because the original gas meter was probably used in another context which was causing the final gasUsed amount to be lower than reality).

fedekunze reviewed (approved):
ACK

yihuang reviewed (approved):
LGTM
